libraries/CryptoInterface/src/CryptoInterface.cpp

```diff
@@ -0,0 +1,566 @@
+    // br_hmac_update(&hmacContext, data, dataLength);
+  
+    // Compute the HMAC output. Assumes message is minimum _ctMinDataLength bytes and maximum _ctMaxDataLength bytes. 
+    // As long as this is true, the correct HMAC output is calculated in constant-time. More constant-time info here: https://www.bearssl.org/constanttime.html
+    // Some extra input bytes are processed, then the output is computed. 
+    // The extra input consists in the dataLength bytes pointed to by data. The dataLength parameter must lie between _ctMinDataLength and _ctMaxDataLength (inclusive); 
+    // _ctMaxDataLength bytes are actually read from data (indicating each data byte can be read multiple times, if dataLength < _ctMaxDataLength). 
+    // Computing time (and memory access pattern) will not depend upon the data byte contents or the value of dataLength.
+    // The output is written in the resultArray buffer, that MUST be large enough to receive it.
+    // The difference _ctMaxDataLength - _ctMinDataLength MUST be less than 2^30 (i.e. about one gigabyte).
+    // This function computes the output properly only if the underlying hash function uses MD padding (i.e. MD5, SHA-1, SHA-224, SHA-256, SHA-384 or SHA-512).
+    // The provided context is NOT modified.
+    br_hmac_outCT(&hmacContext, data, dataLength, _ctMinDataLength, _ctMaxDataLength, resultArray); // returns size_t outputLength
+  
+    return resultArray;
+  }
+  
+  String createBearsslHmacCT(const br_hash_class *hashType, const uint8_t hashTypeNaturalLength, const String &message, const void *hashKey, const size_t hashKeyLength, const size_t hmacLength)
+  {
+    assert(1 <= hmacLength && hmacLength <= hashTypeNaturalLength);
+    
+    uint8_t hmac[hmacLength];
+    createBearsslHmacCT(hashType, message.c_str(), message.length(), hashKey, hashKeyLength, hmac, hmacLength);
+    return TypeCast::uint8ArrayToHexString(hmac, hmacLength);
+  }
+}
+
+namespace CryptoInterface 
+{
+  void setCtMinDataLength(const size_t ctMinDataLength) 
+  {
+    assert(ctMaxDataLength() - ctMinDataLength <= CT_MAX_DIFF);
+    _ctMinDataLength = ctMinDataLength; 
+  }
+  size_t ctMinDataLength() {return _ctMinDataLength;}
+
+  void setCtMaxDataLength(const size_t ctMaxDataLength) 
+  { 
+    assert(ctMaxDataLength - ctMinDataLength() <= CT_MAX_DIFF);
+    _ctMaxDataLength = ctMaxDataLength; 
+  }
+  size_t ctMaxDataLength() {return _ctMaxDataLength;}
+
+  void setWarningsEnabled(bool warningsEnabled) { _warningsEnabled = warningsEnabled; }
+  bool warningsEnabled() { return _warningsEnabled; }
+
+  void setNonceGenerator(nonceGeneratorType nonceGenerator) { _nonceGenerator = nonceGenerator; }
+  nonceGeneratorType getNonceGenerator() { return _nonceGenerator; }
+
+
+  // #################### MD5 ####################
+    
+  // resultArray must have size MD5_NATURAL_LENGTH or greater
+  void *md5Hash(const void *data, const size_t dataLength, void *resultArray)
```

Comment:
We have MD5 stuff in the core, too, which should be migrated to use this (but with same API, probably, so turned into just a wrapper).

Reply:
I think this is still valid. The core would likely make use of this interface for ease of use and for cleaning up current code. Also, crcXX should eventually be integrated here.